### PR TITLE
GHA updates

### PR DESCRIPTION
**Pull request by jamessan**
* Update to clang-12 for ASAN/TSAN testing
* Fix caching for Windows builds
* Grant write permission to "publish" phase of release workflow

### CODE REVIEW

**Review by mhinz — APPROVED** [review]

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -137,6 +137,8 @@ jobs:
   publish:
     needs: [linux, appimage, macOS, windows]
     runs-on: ubuntu-20.04
+    permissions:
+      contents: write
```
Comment (blueyed) [inline .github/workflows/release.yml:141]: What does `contents` here refer to / why is it needed?

Comment (jamessan) [reply .github/workflows/release.yml:141]: https://docs.github.com/en/rest/reference/permissions-required-for-github-apps#permission-on-contents gives the details of what are allowed.

I switched all of the neovim org's actions to read-only access by default, so we now need to grant specific privileges for the default token (or create our own).

Comment (blueyed) [reply .github/workflows/release.yml:141]: > I switched all of the neovim org's actions to read-only access by default

Where/how?

> we now need to grant specific privileges for the default token (or create our own).

Isn't the default for contents "read" anyway (from the link I've posted above)?

Comment (mhinz) [reply .github/workflows/release.yml:141]: > Where/how?

https://github.com/neovim/neovim/settings/actions

> Isn't the default for contents "read" anyway (from the link I've posted above)?

Nope, almost all scopes are `read/write` by default:

https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token

And important here to note:

_When the `permissions` key is used, all unspecified permissions are set to no access, with the exception of the metadata scope, which always gets read access._

**Pull request merged**